python/rpdk/java/templates/init/guided_aws/TagHelper.java

```diff
@@ -106,32 +85,55 @@ public final boolean shouldUpdateTags(final ResourceModel resourceModel, final R
      * getPreviouslyAttachedTags
      *
      * If stack tags and resource tags are not merged together in Configuration class,
-     * we will get previous attached user defined tags from both handlerRequest.getPreviousResourceTags (stack tags)
+     * we will get previously attached system (with `aws:cloudformation` prefix) and user defined tags
+     * from handlerRequest.getPreviousSystemTags(),
+     * handlerRequest.getPreviousResourceTags (stack tags),
      * and handlerRequest.getPreviousResourceState (resource tags).
+     * System tags can change on resource update if the resource is imported to the stack.
      */
     public Map<String, String> getPreviouslyAttachedTags(final ResourceHandlerRequest<ResourceModel> handlerRequest) {
+        final Map<String, String> previousTags = new HashMap<>();
+
+        // get previous system tags
+        if (handlerRequest.getPreviousSystemTags() != null) {
+            previousTags.putAll(handlerRequest.getPreviousSystemTags());
+        }
+
         // get previous stack level tags from handlerRequest
-        final Map<String, String> previousTags = handlerRequest.getPreviousResourceTags() != null ?
-            handlerRequest.getPreviousResourceTags() : Collections.emptyMap();
+        if (handlerRequest.getPreviousResourceTags() != null) {
+            previousTags.putAll(handlerRequest.getPreviousResourceTags());
+        }
 
         // TODO: get resource level tags from previous resource state based on your tag property name
-        // TODO: previousTags.putAll(handlerRequest.getPreviousResourceState().getTags());
+        // TODO:
+        // previousTags.putAll(handlerRequest.getPreviousResourceState().getTags()); // if tags are not null
         return previousTags;
     }
 
     /**
      * getNewDesiredTags
      *
      * If stack tags and resource tags are not merged together in Configuration class,
-     * we will get new user defined tags from both resource model and previous stack tags.
+     * we will get new desired attached system (with `aws:cloudformation` prefix) and user defined tags
+     * from handlerRequest.getSystemTags(), handlerRequest.getPreviousResourceTags (stack tags),
+     * and handlerRequest.getPreviousResourceState (resource tags).
+     * System tags can change on resource update if the resource is imported to the stack.
      */
     public Map<String, String> getNewDesiredTags(final ResourceModel resourceModel, final ResourceHandlerRequest<ResourceModel> handlerRequest) {
-        // get new stack level tags from handlerRequest
-        final Map<String, String> desiredTags = handlerRequest.getDesiredResourceTags() != null ?
-            handlerRequest.getDesiredResourceTags() : Collections.emptyMap();
+        final Map<String, String> desiredTags = new HashMap<>();
```

Comment:
It's a good point that we don't have to impose system tags because they are an optional feature. I'll add a comment here and in `getPreviouslyAttachedTags` similar to one we had in `generateTagsForCreate`.  We should still have system tags here otherwise they are not added on update after import, if they are enabled.